python/cudf_polars/cudf_polars/experimental/parallel.py

```diff
@@ -0,0 +1,152 @@
+    """
+
+    __slots__ = ("count",)
+
+    def __init__(self, count: int):
+        self.count = count
+
+
+# The hash of an IR object must always map to a
+# unique PartitionInfo object, and we can cache
+# this mapping until evaluation is complete.
+_IR_PARTS_CACHE: MutableMapping[int, PartitionInfo] = {}
+
+
+def _clear_parts_info_cache() -> None:
+    """Clear cached partitioning information."""
+    _IR_PARTS_CACHE.clear()
+
+
+def get_key_name(node: Node | NamedExpr) -> str:
+    """Generate the key name for a Node."""
+    if isinstance(node, NamedExpr):
+        return f"named-{get_key_name(node.value)}"  # pragma: no cover
+    return f"{type(node).__name__.lower()}-{hash(node)}"
+
+
+@singledispatch
+def lower_ir_node(ir: IR, rec) -> IR:
+    """Rewrite an IR node with proper partitioning."""
+    # Return same node by default
+    return reuse_if_unchanged(ir, rec)
+
+
+def lower_ir_graph(ir: IR) -> IR:
+    """Rewrite an IR graph with proper partitioning."""
+    from cudf_polars.dsl.traversal import CachingVisitor
+
+    mapper = CachingVisitor(lower_ir_node)
+    return mapper(ir)
+
+
+def _default_ir_parts_info(ir: IR) -> PartitionInfo:
+    # Single-partition default behavior.
+    # This is used by `_ir_parts_info` for all unregistered IR sub-types.
+    count = max((ir_parts_info(child).count for child in ir.children), default=1)
+    if count > 1:
+        raise NotImplementedError(
+            f"Class {type(ir)} does not support multiple partitions."
+        )  # pragma: no cover
+    return PartitionInfo(count=count)
+
+
+@singledispatch
+def _ir_parts_info(ir: IR) -> PartitionInfo:
+    """IR partitioning-info dispatch."""
+    return _default_ir_parts_info(ir)
+
+
+def ir_parts_info(ir: IR) -> PartitionInfo:
+    """Return the partitioning info for an IR node."""
+    key = hash(ir)
+    try:
+        return _IR_PARTS_CACHE[key]
+    except KeyError:
+        _IR_PARTS_CACHE[key] = _ir_parts_info(ir)
+        return _IR_PARTS_CACHE[key]
+
+
+def _default_ir_tasks(ir: IR) -> MutableMapping[Any, Any]:
+    # Single-partition default behavior.
+    # This is used by `generate_ir_tasks` for all unregistered IR sub-types.
+    if ir_parts_info(ir).count > 1:
+        raise NotImplementedError(
+            f"Failed to generate multiple output tasks for {ir}."
+        )  # pragma: no cover
+
+    child_names = []
+    for child in ir.children:
+        child_names.append(get_key_name(child))
+        if ir_parts_info(child).count > 1:
+            raise NotImplementedError(
+                f"Failed to generate tasks for {ir} with child {child}."
+            )  # pragma: no cover
+
+    key_name = get_key_name(ir)
+    return {
+        (key_name, 0): (
+            ir.do_evaluate,
+            *ir._non_child_args,
+            *((child_name, 0) for child_name in child_names),
+        )
+    }
+
+
+@singledispatch
+def generate_ir_tasks(ir: IR) -> MutableMapping[Any, Any]:
```

Comment:
It's actually very convenient to be able to mutate the graph that is returned by an IR node to construct the global task graph (we are doing this now). It would be tricky to make mypy happy with this change.